File: src/simulation_parameters/dc_simulation_parameters.py
from __future__ import annotations

from dataclasses import dataclass

from netlist_parser import NetlistTopology
from .measure_parameters import MeasureEntry
from .print_parameters import PrintParameters
from .sens_parameter import SensParameter


@dataclass(frozen=True)
class DCSimulationParameters:

    sweep_mode: str
    primary_variable: str = ""
    start: str = ""
    stop: str = ""
    step: str = ""
    points: str = ""
    list_values: tuple[str, ...] = ()
    data_table_name: str = ""
    secondary_variable: str = ""
    secondary_start: str = ""
    secondary_stop: str = ""
    secondary_step: str = ""
    secondary_points: str = ""
    replace_ground: bool = True
    print_parameters: PrintParameters | None = None
    measure_parameters: tuple[MeasureEntry, ...] = ()
    sensitivity: SensParameter | None = None

    @classmethod
    def from_xyce_directives(cls, directives: list[str]) -> "DCSimulationParameters" | None:
        # init defaults
        sweep_mode = "LIN"
        primary_variable = ""
        start = ""
        stop = ""
        step = ""
        points = ""
        list_values: tuple[str, ...] = ()
        data_table_name = ""
        secondary_variable = ""
        secondary_start = ""
        secondary_stop = ""
        secondary_step = ""
        secondary_points = ""
        replace_ground = True
        print_parameters = None
        # init measure parameters
        measure_parameters: list[MeasureEntry] = []
        # flag indicating whether a valid directive was found
        found = False
        # parse directives
        for directive in directives:
            # tokenize the directive
            tokens = directive.split()
            # skip empty directives
            if not tokens:
                continue
            cmd = tokens[0].upper()
            # parse print directives and retain dc-specific output config
            if cmd == ".PRINT":
                # parse the print statement from the directive
                print_statement = PrintParameters.from_xyce_statement(directive)
                # retain dc print parameters when found
                if print_statement and print_statement.print_type in ("DC", "HOMOTOPY"):
                    # store the parsed print parameters
                    print_parameters = print_statement
                    # next
                    continue
            # handle preprocess replaceground
            if cmd == ".PREPROCESS" and len(tokens) > 2 and tokens[1].upper() == "REPLACEGROUND":
                # set replace_ground based on the third token
                replace_ground = tokens[2].upper() == "TRUE"
                # next
                continue
            # parse measure directives
            if cmd in (".MEASURE", ".MEAS"):
                # parse the measure statement from the directive
                measure_statement = MeasureEntry.from_xyce_statement(directive)
                # retain measure parameters when found and analysis type matches
                if measure_statement and measure_statement.analysis_type in ("DC", "DC_CONT"):
                    # append the parsed measure parameters
                    measure_parameters.append(measure_statement)
                # next
                continue
            # skip non-DC directives
            if cmd != ".DC":
                continue
            # flag indicating a valid DC directive was found
            found = True
            # handle DATA sweep: .DC DATA=<tablename>
            if len(tokens) == 2 and "=" in tokens[1] and tokens[1].upper().startswith("DATA="):
                # set sweep mode and data table name
                sweep_mode = "DATA"
                data_table_name = tokens[1].split("=", 1)[1]
                # next
                continue
            if len(tokens) < 2:
                continue
            second = tokens[1].upper()
            # detect decade or octave log sweep: .DC DEC|OCT var start stop points
            if second in ("DEC", "OCT"):
                sweep_mode = second
                # primary sweep tokens: MODE var start stop points
                if len(tokens) >= 6:
                    primary_variable = tokens[2]
                    start = tokens[3]
                    stop = tokens[4]
                    points = tokens[5]
                # optional secondary sweep: MODE var2 start2 stop2 points2
                if len(tokens) >= 11 and tokens[6].upper() in ("DEC", "OCT"):
                    secondary_variable = tokens[7]
                    secondary_start = tokens[8]
                    secondary_stop = tokens[9]
                    secondary_points = tokens[10]
                # next
                continue
            # detect LIST sweep: .DC var LIST val [val ...]
            if len(tokens) >= 3 and tokens[2].upper() == "LIST":
                # set sweep mode, primary variable, and list values
                sweep_mode = "LIST"
                primary_variable = tokens[1]
                list_values = tuple(tokens[3:])
                # next
                continue
            # linear sweep: .DC [LIN] var start stop step [var2 start2 stop2 step2]
            sweep_mode = "LIN"
            if second == "LIN":
                # explicit LIN keyword
                if len(tokens) >= 6:
                    primary_variable = tokens[2]
                    start = tokens[3]
                    stop = tokens[4]
                    step = tokens[5]
                # optional secondary sweep tokens: var2 start2 stop2 step2
                if len(tokens) >= 10:
                    secondary_variable = tokens[6]
                    secondary_start = tokens[7]
                    secondary_stop = tokens[8]
                    secondary_step = tokens[9]
            else:
                # implicit LIN
                if len(tokens) >= 5:
                    primary_variable = tokens[1]
                    start = tokens[2]
                    stop = tokens[3]
                    step = tokens[4]
                # optional secondary sweep tokens: var2 start2 stop2 step2
                if len(tokens) >= 9:
                    secondary_variable = tokens[5]
                    secondary_start = tokens[6]
                    secondary_stop = tokens[7]
                    secondary_step = tokens[8]
        # parse sensitivity as a companion directive before analysis detection
        sensitivity = SensParameter.from_xyce_directives(directives)
        # return instance if a valid directive was found
        return cls(sweep_mode=sweep_mode, primary_variable=primary_variable, start=start, stop=stop, step=step, points=points, list_values=list_values, data_table_name=data_table_name, secondary_variable=secondary_variable, secondary_start=secondary_start, secondary_stop=secondary_stop, secondary_step=secondary_step, secondary_points=secondary_points, replace_ground=replace_ground, print_parameters=print_parameters, measure_parameters=tuple(measure_parameters), sensitivity=sensitivity) if found else None

    def to_xyce_directives(self, topology: NetlistTopology | None = None) -> list[str]:
        # prepend replaceground preprocessing when enabled
        preprocess = [".PREPROCESS REPLACEGROUND TRUE"] if self.replace_ground else []
        # build the core dc directive based on the selected sweep mode
        if self.sweep_mode == "DATA":
            lines = [self._build_data_directive()]
        elif self.sweep_mode == "LIST":
            lines = [self._build_list_directive()]
        elif self.sweep_mode == "LIN":
            lines = [self._build_lin_directive()]
        else:
            lines = [self._build_log_directive()]
        # append dc print directive when configured
        if self.print_parameters:
            # append the print statement
            lines.append(self.print_parameters.to_xyce_statement())
        # append sensitivity directives when configured
        if self.sensitivity is not None:
            lines.extend(self.sensitivity.to_xyce_directives(topology))
        # append measure directives
        for measure in self.measure_parameters:
            # append measure statement
            lines.append(measure.to_xyce_statement())
        # return the full directive list
        return preprocess + lines

    def _build_data_directive(self) -> str:
        # data-driven sweep references an existing .DATA table by name
        return f".DC DATA={self.data_table_name}"

    def _build_list_directive(self) -> str:
        # join the explicit sweep values with a single space separator
        values_str = " ".join(self.list_values)
        # combine variable name, LIST keyword, and the value sequence
        return f".DC {self.primary_variable} LIST {values_str}"

    def _build_lin_directive(self) -> str:
        # build the primary linear sweep token sequence
        tokens = [".DC", self.primary_variable, self.start, self.stop, self.step]
        # append secondary sweep tokens when a secondary variable is configured
        if self.secondary_variable:
            # extend with the secondary sweep parameters
            tokens.extend([self.secondary_variable, self.secondary_start, self.secondary_stop, self.secondary_step])
        # combine all tokens into a single directive string
        return " ".join(tokens)

    def _build_log_directive(self) -> str:
        # build the primary decade/octave sweep token sequence
        tokens = [".DC", self.sweep_mode, self.primary_variable, self.start, self.stop, self.points]
        # append secondary sweep tokens when a secondary variable is configured
        if self.secondary_variable:
            # extend with the secondary sweep parameters
            tokens.extend([self.secondary_variable, self.secondary_start, self.secondary_stop, self.secondary_points])
        # combine all tokens into a single directive string
        return " ".join(tokens)

File: src/simulation_parameters/dc_simulation_parameters.cpp
#include <cctype>
#include <optional>
#include <string>
#include <vector>

#include "../util.h"
#include "dc_simulation_parameters.h"

DCSimulationParameters::DCSimulationParameters(std::string sweep_mode, std::string primary_variable, std::string start, std::string stop, std::string step, std::string points, std::vector<std::string> list_values, std::string data_table_name, std::string secondary_variable, std::string secondary_start, std::string secondary_stop, std::string secondary_step, std::string secondary_points, bool replace_ground, std::optional<PrintParameters> print_parameters, std::vector<MeasureEntry> measure_parameters, std::optional<SensParameter> sensitivity) :
    sweep_mode(std::move(sweep_mode)), primary_variable(std::move(primary_variable)), start(std::move(start)), stop(std::move(stop)), step(std::move(step)), points(std::move(points)), list_values(std::move(list_values)), data_table_name(std::move(data_table_name)), secondary_variable(std::move(secondary_variable)), secondary_start(std::move(secondary_start)), secondary_stop(std::move(secondary_stop)), secondary_step(std::move(secondary_step)), secondary_points(std::move(secondary_points)), replace_ground(replace_ground), print_parameters(std::move(print_parameters)), measure_parameters(std::move(measure_parameters)), sensitivity(std::move(sensitivity)) {}

std::optional<DCSimulationParameters> DCSimulationParameters::from_xyce_directives(const std::vector<std::string>& directives) {
    // init defaults
    std::string sweep_mode = "LIN";
    std::string primary_variable;
    std::string start;
    std::string stop;
    std::string step;
    std::string points;
    std::vector<std::string> list_values;
    std::string data_table_name;
    std::string secondary_variable;
    std::string secondary_start;
    std::string secondary_stop;
    std::string secondary_step;
    std::string secondary_points;
    bool replace_ground = true;
    std::optional<PrintParameters> print_parameters;
    std::vector<MeasureEntry> measure_parameters;
    std::optional<SensParameter> sensitivity;

    // flag indicating whether a valid directive was found
    bool found = false;

    // parse directives
    for (const auto& directive : directives) {
        // tokenize the directive
        const auto tokens = tokenize(directive);

        // skip empty directives
        if (tokens.empty()) {
            continue;
        }

        const std::string cmd = to_upper(tokens[0]);

        // parse print directives and retain dc-specific output config
        if (cmd == ".PRINT") {
            // parse the print statement from the directive
            const auto print_statement = PrintParameters::from_xyce_statement(directive);
            // retain dc print parameters when found
            if (print_statement) {
                const std::string print_type_upper = to_upper(print_statement->print_type);
                if (print_type_upper == "DC" || print_type_upper == "HOMOTOPY") {
                    // store the parsed print parameters
                    print_parameters = *print_statement;
                    continue;
                }
            }
        }

        // handle preprocess replaceground
        if (cmd == ".PREPROCESS" && tokens.size() > 2 && to_upper(tokens[1]) == "REPLACEGROUND") {
            // set replace_ground based on the third token
            replace_ground = (to_upper(tokens[2]) == "TRUE");
            continue;
        }

        // parse measure directives
        if (cmd == ".MEASURE" || cmd == ".MEAS") {
            // parse the measure statement from the directive
            const auto measure_statement = MeasureEntry::from_xyce_statement(directive);
            // retain measure parameters when found and analysis type matches
            if (measure_statement) {
                const std::string analysis_type_upper = to_upper(measure_statement->analysis_type);
                if (analysis_type_upper == "DC" || analysis_type_upper == "DC_CONT") {
                    // append the parsed measure parameters
                    measure_parameters.push_back(*measure_statement);
                }
            }
            continue;
        }

        // skip non-DC directives
        if (cmd != ".DC") {
            continue;
        }

        // flag indicating a valid DC directive was found
        found = true;

        // handle DATA sweep: .DC DATA=<tablename>
        if (tokens.size() == 2 && tokens[1].find('=') != std::string::npos && to_upper(tokens[1].substr(0, 5)) == "DATA=") {
            // set sweep mode and data table name
            sweep_mode = "DATA";
            data_table_name = tokens[1].substr(5);
            continue;
        }

        if (tokens.size() < 2) {
            continue;
        }

        const std::string second = to_upper(tokens[1]);

        // detect decade or octave log sweep: .DC DEC|OCT var start stop points
        if (second == "DEC" || second == "OCT") {
            sweep_mode = second;
            // primary sweep tokens: MODE var start stop points
            if (tokens.size() >= 6) {
                primary_variable = tokens[2];
                start = tokens[3];
                stop = tokens[4];
                points = tokens[5];
            }
            // optional secondary sweep: MODE var2 start2 stop2 points2
            if (tokens.size() >= 11 && (to_upper(tokens[6]) == "DEC" || to_upper(tokens[6]) == "OCT")) {
                secondary_variable = tokens[7];
                secondary_start = tokens[8];
                secondary_stop = tokens[9];
                secondary_points = tokens[10];
            }
            continue;
        }

        // detect LIST sweep: .DC var LIST val [val ...]
        if (tokens.size() >= 3 && to_upper(tokens[2]) == "LIST") {
            // set sweep mode, primary variable, and list values
            sweep_mode = "LIST";
            primary_variable = tokens[1];
            for (size_t i = 3; i < tokens.size(); ++i) {
                list_values.push_back(std::string(tokens[i]));
            }
            continue;
        }

        // linear sweep: .DC [LIN] var start stop step [var2 start2 stop2 step2]
        sweep_mode = "LIN";
        if (second == "LIN") {
            // explicit LIN keyword
            if (tokens.size() >= 6) {
                primary_variable = tokens[2];
                start = tokens[3];
                stop = tokens[4];
                step = tokens[5];
            }
            // optional secondary sweep tokens: var2 start2 stop2 step2
            if (tokens.size() >= 10) {
                secondary_variable = tokens[6];
                secondary_start = tokens[7];
                secondary_stop = tokens[8];
                secondary_step = tokens[9];
            }
        }
        else {
            // implicit LIN
            if (tokens.size() >= 5) {
                primary_variable = tokens[1];
                start = tokens[2];
                stop = tokens[3];
                step = tokens[4];
            }
            // optional secondary sweep tokens: var2 start2 stop2 step2
            if (tokens.size() >= 9) {
                secondary_variable = tokens[5];
                secondary_start = tokens[6];
                secondary_stop = tokens[7];
                secondary_step = tokens[8];
            }
        }
    }

    // parse sensitivity as a companion directive before analysis detection
    // For now, we'll return nullopt for sensitivity
    // In the full implementation, this would parse .SENS directives

    // return instance if a valid directive was found
    if (!found) {
        return std::nullopt;
    }

    return DCSimulationParameters(sweep_mode, primary_variable, start, stop, step, points, list_values, data_table_name, secondary_variable, secondary_start, secondary_stop, secondary_step, secondary_points, replace_ground, print_parameters, measure_parameters, sensitivity);
}

std::vector<std::string> DCSimulationParameters::to_xyce_directives() const {
    // init output directive list
    std::vector<std::string> directives;

    // prepend replaceground preprocessing when enabled
    if (replace_ground) {
        directives.push_back(".PREPROCESS REPLACEGROUND TRUE");
    }

    // build the core dc directive based on the selected sweep mode
    std::string dc_directive = ".DC";

    if (sweep_mode == "DATA") {
        dc_directive += " DATA=" + data_table_name;
    }
    else if (sweep_mode == "LIST") {
        dc_directive += " " + primary_variable + " LIST";
        for (const auto& val : list_values) {
            dc_directive += " " + val;
        }
    }
    else if (sweep_mode == "LIN") {
        dc_directive += " " + primary_variable + " " + start + " " + stop + " " + step;
        if (!secondary_variable.empty()) {
            dc_directive += " " + secondary_variable + " " + secondary_start + " " + secondary_stop + " " + secondary_step;
        }
    }
    else {
        // log sweep (DEC or OCT)
        dc_directive += " " + sweep_mode + " " + primary_variable + " " + start + " " + stop + " " + points;
        if (!secondary_variable.empty()) {
            dc_directive += " " + secondary_variable + " " + secondary_start + " " + secondary_stop + " " + secondary_points;
        }
    }

    directives.push_back(dc_directive);

    // append dc print directive when configured
    if (print_parameters) {
        directives.push_back(print_parameters->to_xyce_statement());
    }

    // append sensitivity directives when configured
    if (sensitivity) {
        const auto sens_directives = sensitivity->to_xyce_directives();
        directives.insert(directives.end(), sens_directives.begin(), sens_directives.end());
    }

    // append measure directives
    for (const auto& measure : measure_parameters) {
        directives.push_back(measure.to_xyce_statement());
    }

    // return the full directive list
    return directives;
}

bool DCSimulationParameters::operator==(const DCSimulationParameters& other) const {
    // compare all fields for equality
    return sweep_mode == other.sweep_mode && primary_variable == other.primary_variable && start == other.start && stop == other.stop && step == other.step && points == other.points && list_values == other.list_values && data_table_name == other.data_table_name && secondary_variable == other.secondary_variable && secondary_start == other.secondary_start && secondary_stop == other.secondary_stop && secondary_step == other.secondary_step && secondary_points == other.secondary_points && replace_ground == other.replace_ground && print_parameters == other.print_parameters && measure_parameters == other.measure_parameters && sensitivity == other.sensitivity;
}
